storage/src/journal/variable.rs

```diff
@@ -1,2223 +1,1443 @@
+        // Rewind never updates oldest_retained_pos.
+        if size < self.oldest_retained_pos {
+            return Err(Error::InvalidRewind(size));
         }
 
-        // Compute next offset
-        let aligned_offset = compute_next_offset(offset)?;
+        // Read the offset of the first item to discard (at position 'size').
+        let discard_offset = self.offsets.read(size).await?;
+        let discard_section = position_to_section(size, self.items_per_section);
 
-        // If compression is enabled, decompress the item
-        let item = if compressed {
-            let decompressed =
-                decode_all(Cursor::new(&item)).map_err(|_| Error::DecompressionFailed)?;
-            V::decode_cfg(decompressed.as_ref(), cfg).map_err(Error::Codec)?
-        } else {
-            V::decode_cfg(item, cfg).map_err(Error::Codec)?
-        };
+        self.data
+            .rewind_to_offset(discard_section, discard_offset)
+            .await?;
+        self.offsets.rewind(size).await?;
 
-        // Return item
-        Ok((aligned_offset, size as u32, item))
-    }
+        // Update our size
+        self.size = size;
 
-    /// Helper function to read an item from a [Read].
-    async fn read_buffered(
-        reader: &mut Read<Append<E::Blob>>,
-        offset: u32,
-        cfg: &V::Cfg,
-        compressed: bool,
-    ) -> Result<(u32, u64, u32, V), Error> {
-        // Calculate absolute file offset from the item offset
-        let file_offset = offset as u64 * ITEM_ALIGNMENT;
-
-        // If we're not at the right position, seek to it
-        if reader.position() != file_offset {
-            reader.seek_to(file_offset).map_err(Error::Runtime)?;
-        }
+        Ok(())
+    }
 
-        // Read item size (4 bytes)
-        let mut hasher = crc32fast::Hasher::new();
-        let mut size_buf = [0u8; 4];
-        reader
-            .read_exact(&mut size_buf, 4)
-            .await
-            .map_err(Error::Runtime)?;
-        hasher.update(&size_buf);
-
-        // Read remaining
-        let size = u32::from_be_bytes(size_buf) as usize;
-        let buf_size = size.checked_add(4).ok_or(Error::OffsetOverflow)?;
-        let mut buf = vec![0u8; buf_size];
-        reader
-            .read_exact(&mut buf, buf_size)
-            .await
-            .map_err(Error::Runtime)?;
-
-        // Read item
-        let item = &buf[..size];
-        hasher.update(item);
-
-        // Verify integrity
-        let checksum = hasher.finalize();
-        let stored_checksum = u32::from_be_bytes(buf[size..].try_into().unwrap());
-        if checksum != stored_checksum {
-            return Err(Error::ChecksumMismatch(stored_checksum, checksum));
+    /// Append a new item to the journal, returning its position.
+    ///
+    /// The position returned is a stable, consecutively increasing value starting from 0.
+    /// This position is independent of section boundaries and remains constant even after
```

Comment:
Seems weird discussing the implementation behavior here. I'd prefer we completely abstract the concept of "section" away from the user.  Maybe this can be put within the function body?